src/AWS.Deploy.Orchestration/CdkProjectHandler.cs

```diff
@@ -61,18 +63,47 @@ public async Task CreateCdkDeployment(OrchestratorSession session, CloudApplicat
                 needAwsCredentials: true);
         }
 
-        private async Task<string> CreateCdkProjectForDeployment(Recommendation recommendation, OrchestratorSession session)
+        public async Task<string> CreateCdkProjectForDeployment(Recommendation recommendation, OrchestratorSession session, string? saveCdkDirectoryPath = null)
         {
-            var tempDirectoryPath =
+            string? assemblyName;
+            if (string.IsNullOrEmpty(saveCdkDirectoryPath))
+            {
+                saveCdkDirectoryPath =
                 Path.Combine(
                     Constants.CDK.ProjectsDirectory,
                     Path.GetFileNameWithoutExtension(Path.GetRandomFileName()));
-            Directory.CreateDirectory(tempDirectoryPath);
+
+                assemblyName = recommendation.ProjectDefinition.AssemblyName;
+            }
+            else
+            {
+                assemblyName = new DirectoryInfo(saveCdkDirectoryPath).Name;
+            }
+
+            if (string.IsNullOrEmpty(assemblyName))
+                throw new ArgumentNullException("The assembly name for the CDK deployment project cannot be null");
+           
+            Directory.CreateDirectory(saveCdkDirectoryPath);
 
             var templateEngine = new TemplateEngine();
-            await templateEngine.GenerateCDKProjectFromTemplate(recommendation, session, tempDirectoryPath);
+            await templateEngine.GenerateCDKProjectFromTemplate(recommendation, session, saveCdkDirectoryPath, assemblyName);
+
+            GenerateDeploymentRecipeSnapShot(recommendation, saveCdkDirectoryPath);
 
-            return tempDirectoryPath;
+            return saveCdkDirectoryPath;
+        }
+
+        /// <summary>
+        /// Generates a snapshot of the deployment recipe inside the location at which the CDK deployment project is saved.
+        /// </summary>
+        /// <param name="recommendation"><see cref="Recommendation"/></param>
+        /// <param name="saveCdkDirectoryPath"></param>
+        private void GenerateDeploymentRecipeSnapShot(Recommendation recommendation, string saveCdkDirectoryPath)
+        {
+            var sourceFileName = recommendation.Recipe.RecipePath;
+            var recipeId = recommendation.Recipe.Id;
+            var destFileName = Path.Combine(saveCdkDirectoryPath, recipeId + "-" + Guid.NewGuid().ToString() + ".recipe");
+            File.Copy(sourceFileName, destFileName);
```

Comment:
In the recipe definition we need to replace the Id which is where the Guid can come from. We also need to add new fields indicating what recipe id and version this project was saved from.

Reply:
Replaced the RecipeId with a Guid. Also added the ability for customers to provide their own recipe name (and also use a default name if one is not provided).